p2p/pex/addrbook.go

```diff
@@ -295,6 +307,87 @@ func (a *addrBook) GetSelection() []*p2p.NetAddress {
 	return allAddr[:numAddresses]
 }
 
+// GetSelectionWithBias implements AddrBook.
+// It randomly selects some addresses (old & new). Suitable for peer-exchange protocols.
+//
+// Each address is picked randomly from an old or new bucket according to the
+// biasTowardsNewAddrs argument, which must be between [0, 100] (or else is truncated to
+// that range) and determines how biased we are to pick an address from a new
+// bucket.
+func (a *addrBook) GetSelectionWithBias(biasTowardsNewAddrs int) []*p2p.NetAddress {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	if a.size() == 0 {
+		return nil
+	}
+
+	if biasTowardsNewAddrs > 100 {
+		biasTowardsNewAddrs = 100
+	}
+	if biasTowardsNewAddrs < 0 {
+		biasTowardsNewAddrs = 0
+	}
+
+	numAddresses := cmn.MaxInt(
+		cmn.MinInt(minGetSelection, a.size()),
+		a.size()*getSelectionPercent/100)
+	numAddresses = cmn.MinInt(maxGetSelection, numAddresses)
+
+	selection := make([]*p2p.NetAddress, numAddresses)
+
+	oldBucketAndIndex := make(map[int]int)
+	var oldIndex int
+	newBucketAndIndex := make(map[int]int)
+	var newIndex int
+
+	i := 0
+ADDRS_LOOP:
+	for i < numAddresses {
+		pickFromOldBucket := int((float64(i)/float64(numAddresses))*100) >= biasTowardsNewAddrs
+		pickFromOldBucket = (pickFromOldBucket && a.nOld > 0) || a.nNew == 0
+		bucket := make(map[string]*knownAddress)
+
+		// loop until we pick a random non-empty bucket
+		for len(bucket) == 0 {
+			if pickFromOldBucket {
+				oldIndex = a.rand.Intn(len(a.bucketsOld))
+				bucket = a.bucketsOld[oldIndex]
+			} else {
+				newIndex = a.rand.Intn(len(a.bucketsNew))
+				bucket = a.bucketsNew[newIndex]
+			}
+		}
+
+		// pick a random index and loop over the map to return that index
+		randIndex := a.rand.Intn(len(bucket))
+		if pickFromOldBucket {
+			// check if we already added this address
+			if ind, ok := oldBucketAndIndex[oldIndex]; ok && ind == randIndex {
+				continue ADDRS_LOOP
+			}
+			oldBucketAndIndex[oldIndex] = randIndex
+		} else {
+			// check if we already added this address
+			if ind, ok := newBucketAndIndex[newIndex]; ok && ind == randIndex {
+				continue ADDRS_LOOP
+			}
+			newBucketAndIndex[newIndex] = randIndex
+		}
+
+		for _, ka := range bucket {
+			if randIndex == 0 {
+				selection[i] = ka.Addr
+				i++
```

Comment:
can we add more comments here about what's going on, and where this `i` came from and how we're using it